File: Microservice/model-categorizer/node/classifiers/patternMatcher.js
const {
  ProviderOpenAI,
  ProviderAnthropicA,
  ProviderGemini,
  ProviderMeta,
  ProviderMistral,
  SeriesClaude3,
  SeriesClaude2,
  SeriesClaude1,
  TypeO,
  Type35,
  Type4,
  Type45,
  TypeMini,
  TypeOpus,
  TypeSonnet,
  TypeHaiku,
  TypePro,
  TypeFlashLite,
  TypeFlash,
  TypeThinking,
  TypeGemma,
  TypeVision,
  TypeEmbedding,
  CapVision,
  CapFunctionCalling,
  CapChat,
  Version10,
  Version15,
  Version20,
  Version25,
} = require("./constants");

class PatternMatcher {
  constructor() {
    this.providerPatterns = {
      [ProviderOpenAI]: ["openai", "gpt", "o1", "dall-e"],
      [ProviderAnthropicA]: ["anthropic", "claude"],
      [ProviderGemini]: ["gemini", "google"],
      [ProviderMeta]: ["meta", "llama", "meta-llama"],
      [ProviderMistral]: ["mistral", "mixtral"],
    };

    this.seriesPatterns = {
      [SeriesClaude3]: ["claude-3", "claude3", "claude-3.5", "claude-3.7"],
      [SeriesClaude2]: ["claude-2", "claude2"],
      [SeriesClaude1]: ["claude-1", "claude1", "claude-instant"],
      ["Gemini " + Version10]: ["gemini-1.0", "gemini-1.0-pro"],
      ["Gemini " + Version15]: ["gemini-1.5", "gemini-1.5-pro", "gemini-1.5-flash"],
      ["Gemini " + Version20]: ["gemini-2.0", "gemini-2.0-pro", "gemini-2.0-flash"],
      ["Gemini " + Version25]: ["gemini-2.5", "gemini-2.5-pro", "gemini-2.5-flash"],
      "Gemma 2": ["gemma-2"],
      TypeImage: ["dall-e", "imagen", "midjourney", "stable-diffusion"],
      [TypeEmbedding]: ["embedding", "text-embedding", "embed"],
    };

    this.typePatterns = {
      [TypeO]: ["o1", "o3"],
      [Type35]: ["gpt-3.5", "gpt3.5"],
      [Type4]: ["gpt-4", "gpt4", "gpt-4o"],
      [Type45]: ["gpt-4.5", "gpt4.5"],
      [TypeMini]: ["mini"],
      [TypeOpus]: ["opus"],
      [TypeSonnet]: ["sonnet"],
      [TypeHaiku]: ["haiku"],
      [TypePro]: ["pro"],
      [TypeFlashLite]: ["flash-lite"],
      [TypeFlash]: ["flash"],
      [TypeThinking]: ["thinking"],
      [TypeVision]: ["vision", "multimodal"],
      [TypeEmbedding]: ["embedding", "embed", "tts"],
    };

    this.capabilityPatterns = {
      [CapVision]: ["vision", "image", "multimodal"],
      [CapFunctionCalling]: ["function", "tool", "api"],
      [TypeEmbedding]: ["embedding", "embed", "vector"],
      audio: ["whisper", "tts", "speech", "audio"],
      [CapChat]: ["chat", "conversation", "completion"],
    };
  }

  matchProviderByName(providerName) {
    for (const provider of Object.keys(this.providerPatterns)) {
      if (providerName === provider.toLowerCase()) return provider;
    }
    return "";
  }

  matchProviderByPattern(modelName) {
    const modelLower = modelName.toLowerCase();
    for (const [provider, patterns] of Object.entries(this.providerPatterns)) {
      for (const pattern of patterns) {
        if (modelLower.includes(pattern)) return provider;
      }
    }
    return "";
  }

  matchClaudeVersion(modelName) {
    const modelLower = modelName.toLowerCase();
    for (const pattern of this.seriesPatterns[SeriesClaude3]) {
      if (modelLower.includes(pattern)) return SeriesClaude3;
    }
    for (const pattern of this.seriesPatterns[SeriesClaude2]) {
      if (modelLower.includes(pattern)) return SeriesClaude2;
    }
    for (const pattern of this.seriesPatterns[SeriesClaude1]) {
      if (modelLower.includes(pattern)) return SeriesClaude1;
    }
    return "";
  }

  matchGeminiVersion(modelName) {
    const modelLower = modelName.toLowerCase();
    if (modelLower.includes("2.5")) return "Gemini " + Version25;
    if (modelLower.includes("2.0")) return "Gemini " + Version20;
    if (modelLower.includes("1.5")) return "Gemini " + Version15;
    return "Gemini " + Version10;
  }

  matchSeriesByPattern(modelName) {
    const modelLower = modelName.toLowerCase();
    for (const [series, patterns] of Object.entries(this.seriesPatterns)) {
      for (const pattern of patterns) {
        if (modelLower.includes(pattern)) return series;
      }
    }
    return "";
  }

  matchOpenAIType(modelName) {
    const modelLower = modelName.toLowerCase();
    if (modelLower.includes("mini")) return TypeMini;
    if (modelLower.includes("o1") || modelLower.includes("o3")) return TypeO;
    if (modelLower.includes("gpt-4.5") || modelLower.includes("gpt4.5")) return Type45;
    if (modelLower.includes("gpt-4") || modelLower.includes("gpt4")) return Type4;
    if (modelLower.includes("gpt-3.5") || modelLower.includes("gpt3.5")) return Type35;
    return TypeStandard;
  }

  matchAnthropicType(modelName) {
    const m = modelName.toLowerCase();
    if (m.includes("opus")) return TypeOpus;
    if (m.includes("sonnet")) return TypeSonnet;
    if (m.includes("haiku")) return TypeHaiku;
    return TypeStandard;
  }

  matchGeminiType(modelName) {
    const m = modelName.toLowerCase();
    if (m.includes("flash-lite") || m.includes("flash lite")) return TypeFlashLite;
    if (m.includes("thinking")) return TypeThinking;
    if (m.includes("flash")) return TypeFlash;
    if (m.includes("pro")) return TypePro;
    if (m.includes("gemma")) return TypeGemma;
    return TypeStandard;
  }

  matchTypeByPattern(modelName) {
    for (const [type, patterns] of Object.entries(this.typePatterns)) {
      for (const pattern of patterns) {
        if (modelName.includes(pattern)) return type;
      }
    }
    return "";
  }

  matchOpenAIVariant(modelName) {
    const m = modelName.toLowerCase();
    if (m.includes("gpt-4.5")) return "GPT-" + Version45; // match Go mapping
    if (m.includes("gpt-4o-mini")) return "GPT-4o Mini";
    if (m.includes("gpt-4o")) return "GPT-4o";
    if (m.includes("gpt-4-turbo")) return "GPT-4 Turbo";
    if (m.includes("gpt-4-vision")) return "GPT-4 Vision";
    if (m.includes("o1-mini")) return "O1 Mini";
    if (m.includes("o1")) return "O1";
    return "";
  }

  matchAnthropicVariant(modelName) {
    const m = modelName.toLowerCase();
    if (m.includes("claude-3.7")) return "Claude " + Version37;
    if (m.includes("claude-3.5")) return "Claude " + Version35;
    if (m.includes("claude-3")) return "Claude " + Version30;
    if (m.includes("claude-2")) return "Claude " + Version20;
    if (m.includes("claude-instant")) return "Claude Instant";
    return "";
  }

  buildGeminiVariant(modelName) {
    const m = modelName.toLowerCase();
    let version = "";
    if (m.includes("2.5")) version = Version25;
    else if (m.includes("2.0")) version = Version20;
    else if (m.includes("1.5")) version = Version15;
    else if (m.includes("1.0")) version = Version10;

    let type = "";
    if (m.includes("flash-lite") || m.includes("flash lite")) type = TypeFlashLite;
    else if (m.includes("thinking")) type = TypeThinking;
    else if (m.includes("flash")) type = TypeFlash;
    else if (m.includes("pro")) type = TypePro;

    if (version && type) return `Gemini ${version} ${type}`;
    if (version) return `Gemini ${version}`;
    if (type) return `Gemini ${type}`;
    return "";
  }

  addCapabilities(capabilities, modelType, modelName, provider, series) {
    const m = modelName.toLowerCase();
    if (
      m.includes("vision") ||
      m.includes("multimodal") ||
      modelType === Type4 ||
      modelType === Type45 ||
      modelType === TypeO ||
      series === SeriesClaude3 ||
      m.includes("4o") ||
      (series && series.includes("Gemini"))
    ) {
      capabilities[CapVision] = true;
    }

    if (
      modelType === Type4 ||
      modelType === Type45 ||
      modelType === Type35 ||
      modelType === TypeO ||
      series === SeriesClaude3 ||
      (series && series.includes("Gemini"))
    ) {
      capabilities[CapFunctionCalling] = true;
    }
  }
}

module.exports = { PatternMatcher };

File: Deprecated/Microservice/model-categorizer-rust/src/classifiers/pattern_matcher.rs
use std::collections::HashMap;
use super::{
    PROVIDER_OPENAI,
    PROVIDER_ANTHROPIC,
    PROVIDER_GEMINI,
    PROVIDER_META,
    PROVIDER_MISTRAL,
    PROVIDER_OPENROUTER,
    SERIES_CLAUDE3,
    SERIES_CLAUDE2,
    SERIES_CLAUDE1,
    TYPE_IMAGE,
    TYPE_EMBEDDING,
    TYPE_O,
    TYPE_35,
    TYPE_4,
    TYPE_45,
    TYPE_MINI,
    TYPE_OPUS,
    TYPE_SONNET,
    TYPE_HAIKU,
    TYPE_PRO,
    TYPE_GEMMA,
    TYPE_FLASH_LITE,
    TYPE_FLASH,
    TYPE_THINKING,
    TYPE_VISION,
    TYPE_STANDARD,
    CAP_VISION,
    CAP_FUNCTION_CALLING,
    CAP_EMBEDDING,
    CAP_CHAT,
    VERSION_10,
    VERSION_15,
    VERSION_20,
    VERSION_25,
    VERSION_30,
    VERSION_35,
    VERSION_37,
    VERSION_45,
};

/// PatternMatcher handles all pattern-based identification for models
pub struct PatternMatcher {
    pub provider_patterns: HashMap<String, Vec<String>>,
    pub series_patterns: HashMap<String, Vec<String>>,
    pub type_patterns: HashMap<String, Vec<String>>,
    pub capability_patterns: HashMap<String, Vec<String>>,
}

impl PatternMatcher {
    /// Creates a new PatternMatcher with predefined patterns
    pub fn new() -> Self {
        let mut provider_patterns = HashMap::new();
        provider_patterns.insert(PROVIDER_OPENAI.to_string(), vec!["openai".into(), "gpt".into(), "o1".into(), "dall-e".into()]);
        provider_patterns.insert(PROVIDER_ANTHROPIC.to_string(), vec!["anthropic".into(), "claude".into()]);
        provider_patterns.insert(PROVIDER_GEMINI.to_string(), vec!["gemini".into(), "google".into()]);
        provider_patterns.insert(PROVIDER_META.to_string(), vec!["meta".into(), "llama".into(), "meta-llama".into()]);
        provider_patterns.insert(PROVIDER_MISTRAL.to_string(), vec!["mistral".into(), "mixtral".into()]);
        provider_patterns.insert(PROVIDER_OPENROUTER.to_string(), vec!["openrouter".into()]);

        let mut series_patterns = HashMap::new();
        series_patterns.insert(SERIES_CLAUDE3.to_string(), vec!["claude-3".into(), "claude3".into(), "claude-3.5".into(), "claude-3.7".into()]);
        series_patterns.insert(SERIES_CLAUDE2.to_string(), vec!["claude-2".into(), "claude2".into()]);
        series_patterns.insert(SERIES_CLAUDE1.to_string(), vec!["claude-1".into(), "claude1".into(), "claude-instant".into()]);
        series_patterns.insert(format!("{} {}", "Gemini", VERSION_10), vec!["gemini-1.0".into(), "gemini-1.0-pro".into()]);
        series_patterns.insert(format!("{} {}", "Gemini", VERSION_15), vec!["gemini-1.5".into(), "gemini-1.5-pro".into(), "gemini-1.5-flash".into()]);
        series_patterns.insert(format!("{} {}", "Gemini", VERSION_20), vec!["gemini-2.0".into(), "gemini-2.0-pro".into(), "gemini-2.0-flash".into()]);
        series_patterns.insert(format!("{} {}", "Gemini", VERSION_25), vec!["gemini-2.5".into(), "gemini-2.5-pro".into(), "gemini-2.5-flash".into()]);
        series_patterns.insert("Gemma 2".into(), vec!["gemma-2".into()]);
        series_patterns.insert(TYPE_IMAGE.to_string(), vec!["dall-e".into(), "imagen".into(), "midjourney".into(), "stable-diffusion".into()]);
        series_patterns.insert(TYPE_EMBEDDING.to_string(), vec!["embedding".into(), "text-embedding".into(), "embed".into()]);

        let mut type_patterns = HashMap::new();
        type_patterns.insert(TYPE_O.to_string(), vec!["o1".into(), "o3".into(), "o4".into()]);
        type_patterns.insert(TYPE_35.to_string(), vec!["gpt-3.5".into(), "gpt3.5".into()]);
        type_patterns.insert(TYPE_4.to_string(), vec!["gpt-4".into(), "gpt4".into(), "gpt-4o".into()]);
        type_patterns.insert(TYPE_45.to_string(), vec!["gpt-4.5".into(), "gpt4.5".into()]);
        type_patterns.insert(TYPE_MINI.to_string(), vec!["mini".into()]);
        type_patterns.insert(TYPE_OPUS.to_string(), vec!["opus".into()]);
        type_patterns.insert(TYPE_SONNET.to_string(), vec!["sonnet".into()]);
        type_patterns.insert(TYPE_HAIKU.to_string(), vec!["haiku".into()]);
        type_patterns.insert(TYPE_PRO.to_string(), vec!["pro".into()]);
        type_patterns.insert(TYPE_FLASH_LITE.to_string(), vec!["flash-lite".into()]);
        type_patterns.insert(TYPE_FLASH.to_string(), vec!["flash".into()]);
        type_patterns.insert(TYPE_THINKING.to_string(), vec!["thinking".into()]);
        type_patterns.insert(TYPE_VISION.to_string(), vec!["vision".into(), "multimodal".into()]);
        type_patterns.insert(TYPE_EMBEDDING.to_string(), vec!["embedding".into(), "embed".into(), "tts".into()]);

        let mut capability_patterns = HashMap::new();
        capability_patterns.insert(CAP_VISION.to_string(), vec!["vision".into(), "image".into(), "multimodal".into()]);
        capability_patterns.insert(CAP_FUNCTION_CALLING.to_string(), vec!["function".into(), "tool".into(), "api".into()]);
        capability_patterns.insert(CAP_EMBEDDING.to_string(), vec!["embedding".into(), "embed".into(), "vector".into()]);
        capability_patterns.insert("audio".into(), vec!["whisper".into(), "tts".into(), "speech".into(), "audio".into()]);
        capability_patterns.insert(CAP_CHAT.to_string(), vec!["chat".into(), "conversation".into(), "completion".into()]);

        PatternMatcher { provider_patterns, series_patterns, type_patterns, capability_patterns }
    }

    /// Match a provider by exact name
    pub fn match_provider_by_name(&self, provider_name: &str) -> String {
        for provider in self.provider_patterns.keys() {
            if provider_name == provider.to_lowercase() {
                return provider.clone();
            }
        }
        "".into()
    }

    /// Match a provider based on patterns in the model name
    pub fn match_provider_by_pattern(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        for (provider, patterns) in &self.provider_patterns {
            for pat in patterns {
                if lower.contains(pat) {
                    return provider.clone();
                }
            }
        }
        "".into()
    }

    /// Match Claude series versions
    pub fn match_claude_version(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        for pat in &self.series_patterns[SERIES_CLAUDE3] { if lower.contains(pat) { return SERIES_CLAUDE3.into() } }
        for pat in &self.series_patterns[SERIES_CLAUDE2] { if lower.contains(pat) { return SERIES_CLAUDE2.into() } }
        for pat in &self.series_patterns[SERIES_CLAUDE1] { if lower.contains(pat) { return SERIES_CLAUDE1.into() } }
        "".into()
    }

    /// Match Gemini series versions
    pub fn match_gemini_version(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("2.5") { return format!("{} {}", "Gemini", VERSION_25) }
        if lower.contains("2.0") { return format!("{} {}", "Gemini", VERSION_20) }
        if lower.contains("1.5") { return format!("{} {}", "Gemini", VERSION_15) }
        format!("{} {}", "Gemini", VERSION_10)
    }

    /// Match series by generic patterns
    pub fn match_series_by_pattern(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        for (series, patterns) in &self.series_patterns {
            for pat in patterns {
                if lower.contains(pat) { return series.clone() }
            }
        }
        "".into()
    }

    /// Match OpenAI-specific type
    pub fn match_openai_type(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("mini") { return TYPE_MINI.into() }
        if lower.contains("o1") || lower.contains("o3") { return TYPE_O.into() }
        if lower.contains("gpt-4.5") { return TYPE_45.into() }
        if lower.contains("gpt-4") { return TYPE_4.into() }
        if lower.contains("gpt-3.5") { return TYPE_35.into() }
        TYPE_STANDARD.into()
    }

    /// Match Anthropic-specific type
    pub fn match_anthropic_type(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("opus") { return TYPE_OPUS.into() }
        if lower.contains("sonnet") { return TYPE_SONNET.into() }
        if lower.contains("haiku") { return TYPE_HAIKU.into() }
        TYPE_STANDARD.into()
    }

    /// Match Gemini-specific type
    pub fn match_gemini_type(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("flash-lite") { return TYPE_FLASH_LITE.into() }
        if lower.contains("thinking") { return TYPE_THINKING.into() }
        if lower.contains("flash") { return TYPE_FLASH.into() }
        if lower.contains("pro") { return TYPE_PRO.into() }
        if lower.contains("gemma") { return TYPE_GEMMA.into() }
        TYPE_STANDARD.into()
    }

    /// Generic type matching by patterns
    pub fn match_type_by_pattern(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        for (typ, patterns) in &self.type_patterns {
            for pat in patterns {
                if lower.contains(pat) { return typ.clone() }
            }
        }
        "".into()
    }

    /// Match OpenAI variant
    pub fn match_openai_variant(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("gpt-4.5") { return format!("GPT-{}", VERSION_45) }
        if lower.contains("gpt-4o-mini") { return "GPT-4o Mini".into() }
        if lower.contains("gpt-4o") { return "GPT-4o".into() }
        if lower.contains("gpt-4-turbo") { return "GPT-4 Turbo".into() }
        if lower.contains("gpt-4-vision") { return "GPT-4 Vision".into() }
        if lower.contains("o1-mini") { return "O1 Mini".into() }
        if lower.contains("o1") { return "O1".into() }
        "".into()
    }

    /// Match Anthropic variant
    pub fn match_anthropic_variant(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        if lower.contains("claude-3.7") { return format!("Claude {}", VERSION_37) }
        if lower.contains("claude-3.5") { return format!("Claude {}", VERSION_35) }
        if lower.contains("claude-3") { return format!("Claude {}", VERSION_30) }
        if lower.contains("claude-2") { return format!("Claude {}", VERSION_20) }
        if lower.contains("claude-instant") { return "Claude Instant".into() }
        "".into()
    }

    /// Build Gemini variant combining version and type
    pub fn build_gemini_variant(&self, model_name: &str) -> String {
        let lower = model_name.to_lowercase();
        let version = if lower.contains("2.5") { VERSION_25 } else if lower.contains("2.0") { VERSION_20 } else if lower.contains("1.5") { VERSION_15 } else if lower.contains("1.0") { VERSION_10 } else { "" }
        .to_string();
        let typ = if lower.contains("flash-lite") { TYPE_FLASH_LITE } else if lower.contains("thinking") { TYPE_THINKING } else if lower.contains("flash") { TYPE_FLASH } else if lower.contains("pro") { TYPE_PRO } else { "" };
        match (version.as_str(), typ) {
            (v, t) if !v.is_empty() && !t.is_empty() => format!("Gemini {} {}", v, t),
            (v, _) if !v.is_empty() => format!("Gemini {}", v),
            (_, t) if !t.is_empty() => format!("Gemini {}", t),
            _ => "".into(),
        }
    }

    /// Add capabilities based on model traits
    pub fn add_capabilities(&self, caps: &mut HashMap<String, bool>, model_type: &str, model_name: &str, provider: &str, series: &str) {
        let lower = model_name.to_lowercase();
        if lower.contains("vision") || lower.contains("multimodal") || [TYPE_4, TYPE_45, TYPE_O].contains(&model_type) || series == SERIES_CLAUDE3 || lower.contains("4o") || series.starts_with("Gemini") {
            caps.insert(CAP_VISION.to_string(), true);
        }
        if [TYPE_4, TYPE_45, TYPE_35, TYPE_O].contains(&model_type) || series == SERIES_CLAUDE3 || series.starts_with("Gemini") {
            caps.insert(CAP_FUNCTION_CALLING.to_string(), true);
        }
    }
} 